File: apple/WordPressTVCore/Sources/WordPressTVCore/Data/WireDTOs.swift
import Foundation

// Wire-format DTOs for the WP.com REST API. These are intentionally `internal`
// and mirror the JSON shape exactly; everything outside this file works with
// domain types. Mapping lives in `Mapping.swift`.
//
// Captured shapes (full reference: companion content-source-spec):
//   GET /rest/v1.1/sites/{site}/posts   -> PostsResponseDTO
//   GET /rest/v1.1/videos/{guid}        -> VideoInfoDTO

// MARK: - Posts (the Latest grid)

struct PostsResponseDTO: Decodable {
    let posts: [PostDTO]
}

struct PostDTO: Decodable {
    let id: Int
    let title: String
    let excerpt: String
    let content: String
    let status: String
    /// Keyed by attachment ID. Usually a single video attachment.
    let attachments: [String: AttachmentDTO]?

    enum CodingKeys: String, CodingKey {
        case id = "ID"
        case title, excerpt, content, status, attachments
    }
}

struct AttachmentDTO: Decodable {
    /// VideoPress GUID, when this attachment is a VideoPress video.
    let videopressGuid: String?
    /// Duration in **seconds**.
    let length: Int?
    let thumbnails: ThumbnailsDTO?

    enum CodingKeys: String, CodingKey {
        case videopressGuid = "videopress_guid"
        case length, thumbnails
    }
}

struct ThumbnailsDTO: Decodable {
    let fmtHd: String?
    let fmtDvd: String?
    let fmtStd: String?

    enum CodingKeys: String, CodingKey {
        case fmtHd = "fmt_hd"
        case fmtDvd = "fmt_dvd"
        case fmtStd = "fmt_std"
    }
}

// MARK: - Video info (playback resolution)

struct VideoInfoDTO: Decodable {
    let guid: String
    let title: String?
    /// Duration in **milliseconds**.
    let duration: Int?
    let poster: String?
    /// Absolute URL of the progressive original. Its directory is the base for
    /// the (filename-only) variant URLs in `files`.
    let original: String
    /// Keyed by rendition name (e.g. `"hd"`); values are bare filenames.
    let files: [String: FileVariantDTO]
}

struct FileVariantDTO: Decodable {
    let mp4: String?
    let hls: String?
    let dash: String?
}

File: shared/src/commonMain/kotlin/com/automattic/wordpresstv/core/data/WireDtos.kt
package com.automattic.wordpresstv.core.data

import kotlinx.serialization.Serializable

// Wire-format DTOs for the WP.com REST API. Intentionally `internal` and mirror
// the JSON shape exactly; everything outside this file works with domain types.
// Mapping lives in `Mapping.kt`.
//
//   GET /wp/v2/sites/{site}/posts        -> List<PostDto>   (bare array)
//   GET /wp/v2/sites/{site}/{taxonomy}   -> List<TermDto>   (taxonomy term id)
//   GET /rest/v1.1/videos/{guid}         -> VideoInfoDto
//
// The posts feed uses the modern wp/v2 endpoint because it filters by the
// custom `language`/`event`/category taxonomies. wp/v2 wraps text fields in
// `{ "rendered": ... }` and carries no attachment block, so VideoPress metadata
// comes from rendered content and the v1.1 video-info endpoint.

// --- Posts (the Latest grid) ---

@Serializable
internal data class PostDto(
    val id: Long,
    val status: String = "",
    val title: RenderedDto = RenderedDto(),
    val excerpt: RenderedDto = RenderedDto(),
    val content: RenderedDto = RenderedDto(),
)

@Serializable
internal data class RenderedDto(val rendered: String = "")

@Serializable
internal data class TermDto(
    val id: Long,
    val name: String = "",
    val slug: String = "",
    val count: Int = 0,
)

// --- Video info (playback resolution) ---

@Serializable
internal data class VideoInfoDto(
    val guid: String = "",
    val title: String? = null,
    /** Duration in **milliseconds**. */
    val duration: Long? = null,
    val poster: String? = null,
    /**
     * Absolute URL of the progressive original. Its directory is the base for
     * the (filename-only) variant URLs in [files].
     */
    val original: String,
    /** Keyed by rendition name (e.g. `"hd"`); values are bare filenames. */
    val files: Map<String, FileVariantDto> = emptyMap(),
)

@Serializable
internal data class FileVariantDto(
    val mp4: String? = null,
    val hls: String? = null,
    val dash: String? = null,
)
